File: gpt.py
from capstone import *
import struct
import common
from array import array
import uuid

# https://en.wikipedia.org/wiki/GUID_Partition_Table
# https://www.google.com/search?q=gpt+partition+table+layout&source=lmns&safe=strict&hl=en&ved=2ahUKEwiXma6V2ernAhXWbc0KHUqnDZYQ_AUoAHoECAEQAA
# https://docs.python.org/3/library/struct.html
# https://en.wikipedia.org/wiki/Logical_block_addressing
# https://developer.apple.com/library/archive/technotes/tn2166/_index.html#//apple_ref/doc/uid/DTS10003927-CH1-SUBSECTION11
# https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_type_GUIDs
# https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding
# https://docs.python.org/3/library/uuid.html

def parse_guid(guid):
    final_guid = ""
    first_bytes = ""
    second_bytes = ""
    third_bytes = ""
    fourth_bytes = ""
    fifth_bytes = ""

    first = struct.unpack("<cccc", guid[:4])
    for byte in first:
        first_bytes += byte.hex()

    second = struct.unpack("<cc", guid[4:6])
    for byte in second:
        second_bytes += byte.hex()
    
    third = struct.unpack("<cc", guid[6:8])
    for byte in third:
        third_bytes += byte.hex()
    
    fourth = struct.unpack(">cc", guid[8:10])
    for byte in fourth:
        fourth_bytes += byte.hex()
    
    fifth = struct.unpack(">cccccc", guid[10:])
    for byte in fifth:
        fifth_bytes += byte.hex()
    
    final_guid = "%s-%s-%s-%s-%s" % (first_bytes, 
                                    second_bytes, 
                                    third_bytes, 
                                    fourth_bytes,
                                    fifth_bytes)
    return final_guid

def parse_flags(flags):
    flag_int = int.from_bytes(flags, "little")
    print("Flag integer: %d" % flag_int)
    platform_req = False
    efi_ign = False
    legacy_bootable = False
    read_only = False
    shadow = False
    hidden = False
    no_drive_letter = False
    successful_boot = False
    is_chrome_part = False
    is_windows_part = False
    rsrv_bits = []
    chrome_retries = []
    chrome_priority = 0

    platform_req = (flag_int & 0)
    efi_ign = (flag_int & 1)
    legacy_bootable = (flag_int & 2)
    read_only = (flag_int & 60)
    shadow = (flag_int & 61)
    hidden = (flag_int & 62)
    no_drive_letter = (flag_int & 63)
    successful_boot = (flag_int & 56)

    if read_only or shadow or hidden or no_drive_letter:
        is_windows_part = True

    for i in range(3,55):
        if flag_int & i:
            rsrv_bits.append(i)
            if i > 47 and i < 56:
                is_chrome_part = True
                if i > 52:
                    chrome_retries.append(i)
                else:
                    chrome_priority = i -48

    print("""Flags: 
    Platform Required: %s
    EFI Ignore: %s
    Legacy BIOS Bootable: %s
    Reserved bits used: """ % (platform_req, efi_ign, legacy_bootable), end="")
    print(rsrv_bits)
    if is_windows_part:
        print("""\t*** WINDOWS PARTITION FLAGS DETECTED ***
    Read-Only: %s
    Shadow Copy: %s
    Hidden Drive: %s
    No Drive Letter (no automount): %s""" %(read_only, shadow, hidden, no_drive_letter))
    if is_chrome_part:
        print("""\t*** CHROME OS PARTITION FLAGS DETECTED ***
    Successful Boot: %s""" % successful_boot)
        print("\tTries remaining: " + str(chrome_retries))
        print("\tBoot Priority: %d" % chrome_priority)

def parse_part(dat):
    part_type_guid = dat[:16]
    part_guid = dat[16:32]
    first_lba = dat[32:40]
    last_lba = dat[40:48]
    flags = dat[48:56]
    name = dat[56:]
    
    if common.hexify(part_type_guid) == "00000000000000000000000000000000":
        return

    print("Partition Name: \"%s\"" % name.decode("utf-16"))

    print("Partition Type GUID: %s" % parse_guid(part_type_guid))
    print("Partition GUID: %s" % parse_guid(part_guid))
    
    first_lba_unpack = struct.unpack("<q", first_lba)
    print("First LBA: %s" % common.hexify(first_lba_unpack))

    last_lba_unpack = struct.unpack("<q", last_lba)
    print("Last LBA: %s" % common.hexify(last_lba_unpack))

    parse_flags(flags)
    print("\n")

def read_gpt(file_name):
    with open(file_name, 'rb') as f:
        _protected_mbr = f.read(512)  # this should all effectively be junk
        part_header = f.read(512)
        sig = part_header[:8]
        rev = part_header[8:12]
        size = part_header[12:16]
        crc32 = part_header[16:20]
        resrvd = part_header[20:24]
        curr_LBA = part_header[24:32]
        backup_LBA = part_header[32:40]
        first_usable = part_header[40:48]
        last_usable = part_header[48:56]
        disk_guid = part_header[56:72]
        starting_LBA = part_header[72:80]
        num_parts = part_header[80:84]
        part_size = part_header[84:88]
        crc32_part_arr = part_header[88:92]

        ctr = 0
        print("GPT Header (not decoded):\n" + '-'*30)
        for byte in part_header:
            if ctr >= 9:
                print("{} ".format(hex(byte)))
                ctr = 0
            else:
                print("{} ".format(hex(byte)), end="")
                ctr += 1
        print()
        struct.unpack("l", sig)


        print("Partition Signiture: %s" % sig.decode("utf-8"))
        
        print("GPT Revision: %s" % common.hexify(rev))

        size_unpack = struct.unpack("<I", size)
        print("Header Size: %s" % (common.hexify(size_unpack)))    
        
        crc32_unpack = struct.unpack("<I", crc32)
        print("Header CRC32: %s" % common.hexify(crc32_unpack)) 
        
        print("Reserved Chunk: %s" % common.hexify(resrvd))
        
        curr_LBA_unpack = struct.unpack("<q", curr_LBA)
        print("Current LBA: %s" % common.hexify(curr_LBA_unpack))
        
        backup_LBA_unpack = struct.unpack("<q", backup_LBA)
        print("Backup LBA: %s" % common.hexify(backup_LBA_unpack))
        
        first_usable_unpack = struct.unpack("<q", first_usable)
        print("First Usable LBA: %s" % common.hexify(first_usable_unpack))
        
        last_usable_unpack = struct.unpack("<q", last_usable)
        print("Last Usable LBA: %s" % common.hexify(last_usable_unpack))
        print("Disk GUID: %s" % parse_guid(disk_guid))
        
        starting_LBA_unpack = struct.unpack("<q", starting_LBA)
        print("Starting LBA of Partition Entries: %s" % common.hexify(starting_LBA_unpack))
        
        num_parts_unpack = struct.unpack("<I", num_parts)
        print("Number of Partitions: %s" % common.hexify(num_parts_unpack))
        
        part_size_unpack = struct.unpack("<I", part_size)
        print("Partition Data Size: %s" % common.hexify(part_size_unpack))

        crc32_part_arr_unpack = struct.unpack("<I", crc32_part_arr)
        print("Partition Table CRC32: %s\n" % common.hexify(crc32_part_arr_unpack))


        for i in range(num_parts_unpack[0]):
            parse_part(f.read(part_size_unpack[0]))

    # Stuff that isnt working: 
    # UUID parsing not returing correct values, even though 
    # it is all correct allignment for rest of structures
    #
    # Stuff on last partition doesnt seem correct 
    # (Detecting both Windows and Chrome OS)

File: rust_test/src/gpt.rs
extern crate capstone;

use std::fs::File;
use std::io::prelude::*;
use std::path::Path;
use bit_field::BitArray;


fn get_hex(dat: &[u8]) -> String {
    let mut out = String::new();
    for x in dat.iter(){
        out.push_str(format!("{:x}", x).as_mut_str());
    }
    out
}

fn parse_guid(dat: &[u8]) -> String {
    let mut out = String::new();

    let first = get_hex(&dat[0..4]);
    let second = get_hex(&dat[4..6]);
    let third = get_hex(&dat[6..8]);
    let fourth = get_hex(&dat[8..10]);
    let fifth = get_hex(&dat[10..]);

    out.push_str(format!("{}-{}-{}-{}-{}", first,
                second, third, fourth, fifth).as_mut_str());

    out
}

fn parse_flags(dat: &[u8]){
    assert_eq!(dat.len(), (56-48));
    
    let mut is_windows_part = false;
    let mut is_chrome_part = false;
    let mut rsrv_bits = Vec::new();
    let mut chrome_retries = Vec::new();
    let mut chrome_priority = 0;

    
    // parse flags
    let platform_req = dat.get_bit(0);
    let efi_ign = dat.get_bit(1);
    let legacy_bootable = dat.get_bit(2);
    let read_only = dat.get_bit(60);
    let shadow = dat.get_bit(61);
    let hidden = dat.get_bit(62);
    let no_drive_letter = dat.get_bit(63);
    let successful_boot = dat.get_bit(56);

    if read_only | shadow | hidden | no_drive_letter{
        is_windows_part = true;
    }

    for i in (3..55){
        if dat.get_bit(i){
            rsrv_bits.push(i as u32);
            if i > 47 && i < 56 {
                if is_windows_part{
                    panic!("Detected both Chrome OS and Windows Partitions");
                } else {
                    is_chrome_part = true;
                }
                if i > 52{
                    chrome_retries.push(i as u32);
                } else {
                    chrome_priority = i - 48;
                }

            }

        }
    }

    println!("Flags:\nPlatform Required: {}\nEFI Ignore: {}",
            platform_req, efi_ign);
    println!("Legacy BIOS Bootable: {}\nReserved bits used: {:?}",
            legacy_bootable, rsrv_bits);
    if is_windows_part{
        println!("Detected Windows Partition (from flags)");
        println!("Read-Only: {}\nShadow Copy: {}, Hidden Drive: {}",
                read_only, shadow, hidden);
        println!("No Drive Letter (no automount): {}", no_drive_letter);
    } else if is_chrome_part {
        println!("Detected Chrome OS Partition (from flags)");
        println!("Successful Boot: {}\nTries remaining: {:?}\nBoot Priority: {}",
                successful_boot, chrome_retries, chrome_priority);
    }

}

/// unpacks an integer from bytes
fn unpack_int(bytes: &[u8]) -> u32{
    let mut out = 0u32;
    let mut byte_shifts = 24;
    println!("Unpacking integer: {:?}", bytes);
    for i in 0..4{
        println!("Byte {}", (bytes[i] as u32) << byte_shifts);
        out = out.wrapping_add((bytes[i] as u32) << byte_shifts);
        byte_shifts = byte_shifts - 8;
    }

    println!("Unpacked data: {}", out);
    out
}

/// unpacks long long integer from bytes
fn unpack_long_long(bytes: &[u8]) -> u64 {
    let mut out = 0u64;
    let mut byte_shifts = 48;
    for i in 0..8{
        out = out.wrapping_add((bytes[i] as u64) << byte_shifts);
        byte_shifts = byte_shifts - 8;
    }

    out
}

/// Parse partiton data contained in `data`
fn parse_partition(data: Vec<u8>){

    // parse the data
    let part_type_guid = parse_guid(&data[0..16]);
    let part_guid = parse_guid(&data[16..32]);
    let first_lba = get_hex(&data[32..40]);
    let last_lba = get_hex(&data[40..48]);
    let name = String::from_utf8(data[56..].to_vec()).unwrap();

    // check if empty partition
    if unpack_int(&data[0..16]) == 0u32{
        return;
    }

    // print data
    println!("Partition name: \"{}\"", name);
    println!("Partiton Type GUID: {}", part_type_guid);
    println!("Partition GUID: {}", part_guid);
    println!("First LBA: {}", first_lba);
    println!("Last LBA: {}", last_lba);

    // parse the flags for the partition
    parse_flags(&data[48..56]);
}

/// Read GPT header data
fn read_data(dat: &Vec<u8>, mut file: File){
    let _protected_mbr = &dat[0..512];
    let part_header = dat[512..1024].to_vec();

    // parse the disk's signature into a UTF-8 string
    let sig = match String::from_utf8(part_header[0..8].to_vec()){
        Err(e) => panic!("Failed to convert partition signature to string: {}", e),
        Ok(st) => st
    };

    // get the hex of all of the different things
    let rev = get_hex(&part_header[8..12]); // Parse GPT revision 
    let size = unpack_int(&part_header[12..16]); // parse GPT header size
    let crc32 = get_hex(&part_header[16..20]); // GPT header CRC32 checksum
    let reserved = get_hex(&part_header[20..24]); // Reserved chunk
    let current_lba = get_hex(&part_header[24..32]); // Current Logical Block Address
    let backup_lba = get_hex(&part_header[32..40]); // Backup Logical Block Address
    let first_usable = get_hex(&part_header[40..48]); // First usable sector
    let last_usable = get_hex(&part_header[48..56]); // Last usable sector
    let disk_guid = parse_guid(&part_header[56..72]); // Disk's GUID
    let starting_lba = get_hex(&part_header[72..80]); // Starting Logical Block Address
    let num_parts = unpack_int(&part_header[80..84]); // Number of partitions in table
    let part_size = unpack_int(&part_header[84..88]); // partition data size
    let crc32_part_arr = get_hex(&part_header[88..92]); // Partition table's CRC32

    // print the data
    println!("Partition Signiture: {}", sig);
    println!("GPT Revision: {}", rev);
    println!("Table CRC32: {}", crc32);
    println!("Reserved chunk: {}", reserved);
    println!("Current LBA: {}", current_lba);
    println!("Backup LBA: {}", backup_lba);
    println!("First usable sector: {}", first_usable);
    println!("Last usable sector: {}", last_usable);
    println!("Disk GUID: {}", disk_guid);
    println!("Starting LBA: {}", starting_lba);
    println!("Partition data size: {}", part_size);
    println!("Partition table CRC32: {}", crc32_part_arr);

    // loop over each partition and read its data
    for i in 0..num_parts {
        let mut part_data = vec![0; part_size as usize];
        file.read_exact(part_data.as_mut_slice()).unwrap();
        parse_partition(part_data);
    }

}

pub fn load_and_read(path: &Path){
    // Open the file from `path`
    let mut file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", path.display(), why),
        Ok(file) => file,
    };

    // Read the file contents into a string, returns `io::Result<usize>`
    let mut v = vec![0; 1024];
    match file.read(&mut v) {
        Err(why) => panic!("couldn't read {}: {}", path.display(), why),
        Ok(size) => size,
    };

    read_data(&v, file)

}